File: Python/ByTier/F/2034_F1_Khayyam_s_Royal_Decree_Easy_Version.py
# Problem: CF 2034 F1 - Khayyam's Royal Decree (Easy Version)
# https://codeforces.com/contest/2034/problem/F1

"""
Algorithm: Dynamic Programming with Combinatorics
Purpose: Calculate expected value of gems in satchel after drawing from a chest with special conditions.

Time Complexity:
    - Preprocessing factorial and inverse factorial: O(N)
    - For each test case:
        - Sorting points: O(k log k)
        - Processing all points: O(k^2)
        - Final computation: O(k)
    Overall: O(N + t * k^2) where N = 400002, t is number of test cases, k <= 500.
Space Complexity:
    - Arrays for factorials and inverses: O(N)
    - Additional arrays for processing points: O(k)

Note: MOD = 998244353 (Prime), used for modular arithmetic.
"""

ans = []
MOD = 0x3B800001  # This is 998244353 in decimal
N = 400_002
fact = [1]  # Factorial array modulo MOD
for i in range(1, N):
    fact.append(fact[-1] * i % MOD)

# Precompute modular inverses of factorials
anti = [0] * N
anti[-1] = pow(fact[-1], -1, MOD)  # Modular inverse of last factorial
for i in range(N - 2, -1, -1):
    anti[i] = anti[i + 1] * (i + 1) % MOD


# Combination function C(n, k) mod MOD using precomputed factorials
def cmb(n, k):
    if not 0 <= k <= n:
        return 0
    return fact[n] * anti[n - k] * anti[k] % MOD


for _ in range(int(input())):
    n, m, k = [int(t) for t in input().split()]
    pts = [[int(t) for t in input().split()] for _ in range(k)]
    pts.sort(key=lambda x: x[0] + x[1])  # Sort points based on sum of coordinates
    v = [0] * k

    # Dynamic programming to compute contribution of each scroll condition
    for i in range(k):
        a, b = pts[i]
        S = cmb(a + b, b)  # Base count of paths reaching (a,b)
        for j in range(i):
            c, d = pts[j]
            if c <= a and d <= b:
                # Add contribution from previous state reachable to (a,b)
                S += cmb(a + b - c - d, a - c) * v[j]
        v[i] = S % MOD

    tot = 0
    # Final computation for expected value using all scroll conditions
    for i in range(k):
        a, b = pts[i]
        tot += cmb(n - a + m - b, n - a) * v[i] * (2 * (n - a) + m - b)

    tot = tot * pow(cmb(n + m, m), -1, MOD) % MOD  # Normalize by total combinations
    tot += 2 * n + m
    ans.append(tot % MOD)

for w in ans:
    print(w)


# https://github.com/VaHiX/codeForces/

File: CPP/ByRound/2034/2034_F1_Khayyam_s_Royal_Decree_Easy_Version.cpp
// Problem: CF 2034 F1 - Khayyam's Royal Decree (Easy Version)
// https://codeforces.com/contest/2034/problem/F1

/*
 * Problem: Khayyam's Royal Decree (Easy Version)
 * 
 * Purpose: 
 *   This code computes the expected value of Khayyam's satchel at the end of the game,
 *   where gems are drawn from a chest and special configurations trigger royal decrees
 *   that double the value of the satchel.
 *
 * Algorithms/Techniques:
 *   - Dynamic Programming with inclusion-exclusion principle
 *   - Modular arithmetic with precomputed factorials and inverses
 *   - Combinatorial calculations for probabilities
 *   - Fast exponentiation for modular inverse
 *
 * Time Complexity:
 *   O(k^2 * (n + m) + k * log(P)) per test case, where P is the modulus 998244353.
 *   Preprocessing of factorials and inverses takes O(n + m).
 *   The DP calculation and final expected value computation are bounded by k^2 and
 *   the number of combinations involved.
 *
 * Space Complexity:
 *   O(n + m + k) for storing factorials, inverses, and auxiliary arrays.
 */

#include <algorithm>
#include <iostream>
#include <utility>

#define int long long
using namespace std;
using Pii = pair<int, int>;
const int N = 5e3 + 5, L = 4e5 + 5, P = 998244353;
int t, n, m, k, f[L], I[L], g[N], r;
Pii a[N];

// Fast exponentiation to compute modular inverse
int Q(int x, int y) {
  if (!y) {
    return 1;
  }
  int z = Q(x, y / 2);
  return z * z % P * (y & 1 ? x : 1) % P;
}

// Computes combination C(x, y) = x! / (y! * (x - y)!)
int C(int x, int y) { return f[x] * I[y] % P * I[x - y] % P; }

signed main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);
  
  // Precompute factorials and their modular inverses
  f[0] = I[0] = 1;
  for (int i = 1; i < L; i++) {
    f[i] = f[i - 1] * i % P;
  }
  I[L - 1] = Q(f[L - 1], P - 2);
  for (int i = L - 2; i; i--) {
    I[i] = I[i + 1] * (i + 1) % P;
  }
  
  for (cin >> t; t--; fill(g, g + k + 1, r = 0)) {
    cin >> n >> m >> k;
    
    // Transform scroll conditions into "remaining gems" format
    for (int i = 1; i <= k; i++) {
      cin >> a[i].first >> a[i].second;
      a[i].first = n - a[i].first;
      a[i].second = m - a[i].second;
    }
    
    // Sort scroll conditions by number of remaining gems
    sort(a + 1, a + k + 1);
    
    // DP: calculate contribution of each scroll condition
    for (int i = 1; i <= k; i++) {
      // Contribution of current scroll condition, factoring in all previous ones
      g[i] = C(a[i].first + a[i].second, a[i].first) *
             (2 * a[i].first + a[i].second) % P;
      
      // Add contributions from earlier conditions using inclusion-exclusion
      for (int j = 0; j < i; j++) {
        if (a[j].second <= a[i].second) {
          g[i] = (g[i] +
                  g[j] * C(a[i].first + a[i].second - a[j].first - a[j].second,
                           a[i].first - a[j].first)) %
                 P;
        }
      }
      
      // Add total contribution to final result
      r = (r + g[i] * C(n - a[i].first + m - a[i].second, n - a[i].first)) % P;
    }
    
    // Final expected value includes base value plus correction for all paths
    cout << (r + (2 * n + m) * C(n + m, n)) % P * Q(C(n + m, n), P - 2) % P
         << '\n';
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/